src/Tuura/Plato/Translation.hs

```diff
@@ -15,12 +19,36 @@ instance Ord Signal
     where
         compare (Signal x) (Signal y) = compare x y
 
-addErrors :: (Eq a, Show a) => [a] -> [a] -> [a] -> String
-addErrors unused incons undef = un ++ ic ++ ud
+addErrors :: (Eq a, Show a) => [a] -> [a] -> [a] -> [a] -> String
+addErrors unused incons undef invInit = un ++ ic ++ ud ++ iv
+  where
+    un = if (unused  /= []) then ("The following signals are not declared as input, output or internal: \n" ++ unlines (map show unused) ++ "\n") else ""
+    ic = if (unused  /= []) then ("The following signals have inconsistent inital states: \n" ++ unlines (map show incons) ++ "\n") else ""
+    ud = if (undef   /= []) then ("The following signals have undefined initial states: \n" ++ unlines (map show undef) ++ "\n") else ""
+    iv = if (invInit /= []) then ("The initial state is an invariant state. Signals with erroneous initial states: \n" ++ unlines (map show invInit) ++ "\n") else ""
+
+validate :: Ord a => [a] -> CircuitConcept a -> ValidationResult a
+validate signs circuit
+    | unused ++ inconsistent ++ undef ++ invInit == [] = Valid
+    | otherwise                             = Invalid unused inconsistent undef invInit
   where
-    un = if (unused /= []) then ("The following signals are not declared as input, output or internal: \n" ++ unlines (map show unused) ++ "\n") else ""
-    ic = if (unused /= []) then ("The following signals have inconsistent inital states: \n" ++ unlines (map show incons) ++ "\n") else ""
-    ud = if (undef  /= []) then ("The following signals have undefined initial states: \n" ++ unlines (map show undef) ++ "\n") else ""
+    unused       = filter ((==Unused) . interface circuit) signs
+    inconsistent = filter ((==Inconsistent) . initial circuit) signs
+    undef        = filter ((==Undefined) . initial circuit) signs
+    invInit      = checkInitialStates signs circuit
+
+checkInitialStates :: Ord a => [a] -> CircuitConcept a -> [a]
+checkInitialStates signs circuit = nubOrd (concatMap (\inv -> if (checkForInvariant inv initialStates) then (invariantError inv) else []) (invariant circuit))
+  where
+    initialStates = map (\s -> Transition { signal = s , newValue = getDefined $ initial circuit s }) signs
+
+invariantError :: [Transition a] -> [a]
+invariantError = map (signal)
+
+checkForInvariant :: Eq a => [Transition a] -> [Transition a] -> Bool
+checkForInvariant invs initialStates
```

Comment:
Rewrite this without explicit recursion. Check out standard `all` and `any` functions in Haskell.